File: backend/go-api/internal/repository/twofa_repo.go
package repository

import (
	"context"
	"errors"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/platform/database"

	"gorm.io/gorm"
)

type twofaRepo struct {
	db *gorm.DB
}

func NewTwoFARepository(db *gorm.DB) TwoFARepository {
	return &twofaRepo{db: db}
}

func (r *twofaRepo) conn(ctx context.Context) *gorm.DB {
	return database.DBFromContext(ctx, r.db)
}

func (r *twofaRepo) CreateToken(ctx context.Context, token *domain.TwoFactorToken) error {
	return r.conn(ctx).Create(token).Error
}

func (r *twofaRepo) FindValidToken(ctx context.Context, userID uint, code string) (*domain.TwoFactorToken, error) {
	var token domain.TwoFactorToken
	if err := r.conn(ctx).Where("user_id = ? AND code = ? AND used = ?", userID, code, false).First(&token).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &token, nil
}

func (r *twofaRepo) InvalidateTokens(ctx context.Context, userID uint) error {
	return r.conn(ctx).Model(&domain.TwoFactorToken{}).
		Where("user_id = ? AND used = ?", userID, false).
		Update("used", true).Error
}

func (r *twofaRepo) FindLatestUnusedToken(ctx context.Context, userID uint) (*domain.TwoFactorToken, error) {
	var token domain.TwoFactorToken
	if err := r.conn(ctx).Where("user_id = ? AND used = ?", userID, false).
		Order("created_at DESC").First(&token).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &token, nil
}

func (r *twofaRepo) IncrementAttempts(ctx context.Context, tokenID uint) error {
	return r.conn(ctx).Model(&domain.TwoFactorToken{}).
		Where("id = ?", tokenID).
		Update("attempts", gorm.Expr("attempts + 1")).Error
}

func (r *twofaRepo) MarkTokenUsed(ctx context.Context, id uint) error {
	return r.conn(ctx).Model(&domain.TwoFactorToken{}).
		Where("id = ?", id).
		Update("used", true).Error
}

func (r *twofaRepo) CreateRecoveryCode(ctx context.Context, code *domain.RecoveryCode) error {
	return r.conn(ctx).Create(code).Error
}

func (r *twofaRepo) FindUnusedRecoveryCodes(ctx context.Context, userID uint) ([]domain.RecoveryCode, error) {
	var codes []domain.RecoveryCode
	if err := r.conn(ctx).Where("user_id = ? AND used = ?", userID, false).Find(&codes).Error; err != nil {
		return nil, err
	}
	return codes, nil
}

func (r *twofaRepo) MarkRecoveryCodeUsed(ctx context.Context, id uint) error {
	return r.conn(ctx).Model(&domain.RecoveryCode{}).
		Where("id = ?", id).
		Updates(map[string]interface{}{"used": true, "used_at": gorm.Expr("NOW()")}).Error
}

File: backend/python-api/app/repository/twofa_repo.py
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.twofa import RecoveryCode, TwoFactorToken


class TwoFARepository:
    def __init__(self, session: AsyncSession):
        self._session = session

    async def create_token(self, token: TwoFactorToken) -> None:
        self._session.add(token)
        await self._session.flush()

    async def find_valid_token(self, user_id: int, code: str) -> TwoFactorToken | None:
        stmt = select(TwoFactorToken).where(
            TwoFactorToken.user_id == user_id,
            TwoFactorToken.code == code,
            TwoFactorToken.used.is_(False),
            TwoFactorToken.expires_at > datetime.now(timezone.utc),
        )
        result = await self._session.execute(stmt)
        return result.scalar_one_or_none()

    async def find_latest_unused_token(self, user_id: int) -> TwoFactorToken | None:
        stmt = (
            select(TwoFactorToken)
            .where(
                TwoFactorToken.user_id == user_id,
                TwoFactorToken.used.is_(False),
            )
            .order_by(TwoFactorToken.created_at.desc())
            .limit(1)
        )
        result = await self._session.execute(stmt)
        return result.scalar_one_or_none()

    async def increment_attempts(self, token_id: int) -> None:
        await self._session.execute(
            update(TwoFactorToken)
            .where(TwoFactorToken.id == token_id)
            .values(attempts=TwoFactorToken.attempts + 1)
        )
        await self._session.flush()

    async def invalidate_tokens(self, user_id: int) -> None:
        await self._session.execute(
            update(TwoFactorToken)
            .where(TwoFactorToken.user_id == user_id, TwoFactorToken.used.is_(False))
            .values(used=True)
        )
        await self._session.flush()

    async def mark_token_used(self, token_id: int) -> None:
        await self._session.execute(
            update(TwoFactorToken).where(TwoFactorToken.id == token_id).values(used=True)
        )
        await self._session.flush()

    # --- Recovery codes ---

    async def create_recovery_code(self, code: RecoveryCode) -> None:
        self._session.add(code)
        await self._session.flush()

    async def find_unused_recovery_codes(self, user_id: int) -> list[RecoveryCode]:
        stmt = select(RecoveryCode).where(
            RecoveryCode.user_id == user_id,
            RecoveryCode.used.is_(False),
        )
        result = await self._session.execute(stmt)
        return list(result.scalars().all())

    async def mark_recovery_code_used(self, code_id: int) -> None:
        await self._session.execute(
            update(RecoveryCode)
            .where(RecoveryCode.id == code_id)
            .values(used=True, used_at=datetime.now(timezone.utc))
        )
        await self._session.flush()
